### PR TITLE
Ammunoz/boolean

### PR DESCRIPTION
**Pull request by ammunoz**

### CODE REVIEW

File: src/Geometry.cpp [hunk 1]
```diff
@@ -55,25 +66,92 @@ Geometry::Geometry( const std::string filename, GLenum r, Shader* s1, Shader* s2
...
+    , program( geo )
+    , programOutline( stencil )
     , vertices( std::move( v ) )
     , colours( std::move( c ) )
-    , normals( {} )
-, ModelMatrix( glm::mat4( 1.0f ) ) {
+    , normals( std::move( n ) )
+    , ModelMatrix( glm::mat4( 1.0f ) ) {
+
+    for ( int i = 0; i < vertices.size() / 3; ++i ) {
```
Comment (exspeed) [inline src/Geometry.cpp]: for ( uint i = 0; i < vertices.size() ; i+=3 ) {
       faces.push_back( i );
       faces.push_back( i + 1 );
      faces.push_back( i + 2 );
   }

File: src/Geometry.cpp [hunk 2]
```diff
@@ -55,25 +66,92 @@ Geometry::Geometry( const std::string filename, GLenum r, Shader* s1, Shader* s2
...
+
+    // Push vertices and colours
+    for ( int i = 0; i < trimesh.n_vertices; ++i ) {
+        glm::vec3 v(
+            trimesh.vertices[3 * i],
+            trimesh.vertices[3 * i + 1],
+            trimesh.vertices[3 * i + 2]
+        );
+        vertices.push_back( v );
+        colours.push_back( glm::vec3( 1.0f, 1.0f, 1.0f ) );
```
Comment (exspeed) [inline src/Geometry.cpp]: you can use emplace here.
vertices.emplace(trimesh.vertices[3 * i], trimesh.vertices[3 * i+1], trimesh.vertices[3 * i+2],)

File: src/Geometry.cpp [hunk 3]
```diff
@@ -55,25 +66,92 @@ Geometry::Geometry( const std::string filename, GLenum r, Shader* s1, Shader* s2
...
+
+        // Calculate normals
+        glm::vec3 l = b - a;
+        glm::vec3 r = c - a;
+        glm::vec3 n = glm::normalize( glm::cross( r, l ) );
+
+        // Push normals, and faces
+        for ( int j = 0; j < 3; ++j ) {
+            normals.push_back( n );
+            faces.push_back( trimesh.triangles[3 * i + j] );
```
Comment (exspeed) [inline src/Geometry.cpp:150]: pushing a face 3 times? is these vertices?

**Comment by DanielDDo** [comment]
quickly went over it, didnt see any glaring problems.

**Review by DanielDDo — APPROVED** [review]

File: src/Geometry.cpp [hunk 4]
```diff
@@ -163,3 +250,19 @@ void Geometry::Destroy() const {
     glDeleteBuffers( 1, &colourBuffer );
     glDeleteBuffers( 1, &normalBuffer );
 }
+
+void Geometry::GetCorkTriMesh( CorkTriMesh& out ) {
+    // Dynamically create copies of geometry data
+    std::vector<GLuint>* nfaces = new std::vector<GLuint>( faces );
```
Comment (exspeed) [inline src/Geometry.cpp]: do we need to delete these later

File: src/Geometry.cpp [hunk 5]
```diff
@@ -163,3 +250,19 @@ void Geometry::Destroy() const {
...
+    std::vector<glm::vec3>* nvertices = new std::vector<glm::vec3>( vertices );
+
+    for ( auto& v : *nvertices ) {
```
Comment (exspeed) [inline src/Geometry.cpp]: ??? is this doing anything

**Review by exspeed — CHANGES_REQUESTED** [review]

File: src/Geometry.cpp (hunk 3 above)
Comment (ammunoz) [reply src/Geometry.cpp:150]: A face F is defined by F = {i, j. k} where i, j, and k are indeces to the different vertex array buffers (i.e., vertices, colours, and normals).

For example:
F = {1, 2, 3} is a face made up of vertices[1], vertices[2], and vertices[3]

Note: This is the case for triangular meshes. For quad mesh, F = {i, j, k, l}. |F| = n for general meshes.

File: src/Geometry.cpp (hunk 4 above)
Comment (ammunoz) [reply src/Geometry.cpp]: Technically yeah but this is temporary and will be abstracted away later. This is just so it's convenient to test with for now.

Also, no where in the code should this be used right now until it's ready

File: src/Geometry.cpp (hunk 5 above)
Comment (ammunoz) [reply src/Geometry.cpp]: Transforming members of nvertices to world space

**Review by exspeed — APPROVED** [review]

**Pull request merged**